src/ledger/LedgerTxnAccountSQL.cpp

```diff
@@ -367,6 +370,348 @@ LedgerTxnRoot::Impl::deleteAccount(LedgerKey const& key)
+        "lastmodified, buyingliabilities, sellingliabilities "
+        ") VALUES ( "
+        ":id, :v1, :v2, :v3, :v4, :v5, :v6, :v7, :v8, :v9, :v10 "
+        ") ON CONFLICT (accountid) DO UPDATE SET "
+        "balance = excluded.balance, "
+        "seqnum = excluded.seqnum, "
+        "numsubentries = excluded.numsubentries, "
+        "inflationdest = excluded.inflationdest, "
+        "homedomain = excluded.homedomain, "
+        "thresholds = excluded.thresholds, "
+        "flags = excluded.flags, "
+        "lastmodified = excluded.lastmodified, "
+        "buyingliabilities = excluded.buyingliabilities, "
+        "sellingliabilities = excluded.sellingliabilities "
+        "WHERE accounts.accountid = excluded.accountid";
+    auto prep = DB.getPreparedStatement(sql);
+    soci::statement& st = prep.statement();
+    st.exchange(soci::use(accountIDs, "id"));
+    st.exchange(soci::use(balances, "v1"));
+    st.exchange(soci::use(seqNums, "v2"));
+    st.exchange(soci::use(subEntryNums, "v3"));
+    st.exchange(soci::use(inflationDests, inflationDestInds, "v4"));
+    st.exchange(soci::use(homeDomains, "v5"));
+    st.exchange(soci::use(thresholds, "v6"));
+    st.exchange(soci::use(flags, "v7"));
+    st.exchange(soci::use(lastModifieds, "v8"));
+    st.exchange(soci::use(buyingLiabilities, liabilitiesInds, "v9"));
+    st.exchange(soci::use(sellingLiabilities, liabilitiesInds, "v10"));
+    st.define_and_bind();
+    {
+        auto timer = DB.getUpsertTimer("account");
+        st.execute(true);
+    }
+}
+
+template <typename T>
+std::string
+marshalToPGArrayItem(PGconn* conn, const T& item)
+{
+    std::ostringstream ss;
+    ss << item;
+    return ss.str();
+}
+
+template <>
+std::string
+marshalToPGArrayItem<std::string>(PGconn* conn, const std::string& item)
+{
+    static const std::string q("\"");
+    char buf[item.size() * 2 + 1];
+    int err = 0;
+    size_t len = PQescapeStringConn(conn, buf, item.c_str(), item.size(), &err);
+    if (err != 0)
+    {
+        throw std::runtime_error("Could not escape string in SQL");
+    }
+    return q + std::string(buf, len) + q;
+}
+
+template <typename T>
+std::string
+marshalToPGArray(PGconn* conn, const std::vector<T>& v,
+                 const std::vector<soci::indicator>* ind = 0)
+{
+    std::string result = "{";
+    for (size_t i = 0; i < v.size(); ++i)
+    {
+        if (i > 0)
+        {
+            result += ", ";
+        }
+        if (ind && (*ind)[i] == soci::i_null)
+        {
+            result += "NULL";
+        }
+        else
+        {
+            result += marshalToPGArrayItem(conn, v[i]);
+        }
+    }
+    result += "}";
+    return result;
+}
+
+static void
+postgresSpecificBulkUpsertAccounts(
+    Database& DB, std::vector<std::string> const& accountIDs,
+    std::vector<int64_t> const& balances, std::vector<int64_t> const& seqNums,
+    std::vector<uint32_t> const& subEntryNums,
```

Comment:
I don't think we can use unsigned types in any SQL interfaces

Reply:
Ok. I'll fiddle with this (it's unsigned in the XDR but, ah right, I forgot we're doing this with the weird signed-with-non-negative-constraint thing in the SQL).

Reply:
fixed.